### ISSUE TITLE
support italic and underline styles

### ISSUE THREAD

**Issue by rhaberkorn** [orbitalquark/scinterm#22]
SCI_STYLESETUNDERLINE and SCI_STYLESETITALIC could be supported via A_UNDERLINE and A_ITALIC.

I might contribute this feature in the future. Right now, I am not highly motivated.

**Comment by orbitalquark** [comment]
Italics might be pretty straightforward, but underlines likely will not be due to https://github.com/orbitalquark/scinterm/blob/bbedf0215c9b0d3049925d3886d983f6d778581d/PlatCurses.h#L9-L22

A patch to Scintilla itself may be required. Here's LexLPeg.cxx that was referenced in that comment: https://github.com/orbitalquark/scintillua/blob/9ee374d2b7c570208d2494d716aec9522af34d38/LexLPeg.cxx#L431-L441

**Comment by rhaberkorn** [comment]
I might not know about the implementation details, but why is A_UNDERLINE harder than A_ITALIC?

**Comment by orbitalquark** [comment]
Sorry, I used the wrong link in my previous comment. I've updated it.

**Comment by orbitalquark** [comment]
Basically, Scintilla does not store underline in font parameters, but Scinterm makes use of font parameters when drawing text. When Scintilla sees that a style has the underline property set, it draws the underline in a separate step, which Scinterm cannot tap into right now.

We would either need to:

- Determine which draw function Scintilla uses for drawing underlines and figure out how to infer text is being underlined, then change the cell attributes to do the underlining for the range.
- Patch Scintilla to allow an override drawing function for underlining text (similar to how there are overrides for DrawWrapMarker and DrawTabArrow).
- Find a way to make use of font weight for storing underline attribute (either patch Scintilla or figure out how to hook into SCI_STYLESETUNDERLINE; see the second link in my original comment).

**Comment by rhaberkorn** [comment]
I see. For my limited needs (SciTECO), I might just overwrite SCI_STYLESETUNDERLINE and SCI_STYLESETITALIC, setting the font weight style internally, since I control the Scintilla message interface. btw. Scinterm could do something similar since you do control the implementation of scintilla_send_message(). This won't help when using the C++ interface, though.

**Comment by orbitalquark** [comment]
You're right, Scinterm does have control over the Scintilla message interface, so option 3 will work. Scintilla has also introduced a new stretch property that Scinterm can make use of, instead of sharing font weight with bold. https://github.com/orbitalquark/scinterm/commit/566097a89fe0b6a1d279f1cb12d071f287146bc9 should implement this. The jinx demo appears to work.

Thanks!

**Comment by rhaberkorn** [comment]
Well of course, it would be nice if we had INDIC_PLAIN as well on Curses. But I guess that's not easy to hack as you would indeed have to implement `SurfaceImpl::LineDraw()` and try to filter out the underlines only.
We cannot directly influence the font attributes for indicators - they appear to be drawn exclusively on top of the regularly rendered fonts.

**Comment by rhaberkorn** [comment]
I am now motivated to get underlines working via indicators as well. Perhaps we can abuse one of the other indicators...

**Comment by rhaberkorn** [comment]
What if we did this:

```patch
diff --git a/PlatCurses.cxx b/PlatCurses.cxx
index df0f9ba..49c1429 100644
--- a/PlatCurses.cxx
+++ b/PlatCurses.cxx
@@ -218,7 +218,13 @@ void SurfaceImpl::RectangleDraw(PRectangle /*rc*/, FillStroke /*fillStroke*/) {}
 
 // Drawing framed rectangles like fold display text, EOL annotations, and INDIC_BOX is not
 // implemented.
-void SurfaceImpl::RectangleFrame(PRectangle /*rc*/, Stroke /*stroke*/) {}
+void SurfaceImpl::RectangleFrame(PRectangle rc, Stroke /*stroke*/) {
+	for (int x = static_cast<int>(std::max(rc.left, clip.left)), y = static_cast<int>(rc.top - 1);
+		x < rc.right; x++) {
+		attr_t attrs = mvwinch(win, y, x) & A_ATTRIBUTES;
+		mvwchgat(win, y, x, 1, attrs | A_UNDERLINE, PAIR_NUMBER(attrs), nullptr);
+	}
+}
 
 // Normally this clears the given portion of the screen with the given background color. In
 // some cases however, it can be determined that whitespace is being drawn. If so, draw it
```

It would allow us to at least use INDIC_BOX to get an underline style. Perhaps the color should also be changed to the stroke color. This would also be useful in spell checking where you might want to highlight words in red and underlines.

However, RectangleFrame() might be called in other situations as well and cause unwanted underline attributes.

**Comment by rhaberkorn** [comment]
Alternatively, an alpha value of 0 could cause underlines and coloring with INDIC_STRAIGHTBOX:

```patch
diff --git a/PlatCurses.cxx b/PlatCurses.cxx
index df0f9ba..e43568e 100644
--- a/PlatCurses.cxx
+++ b/PlatCurses.cxx
@@ -266,9 +266,12 @@ void SurfaceImpl::AlphaRectangle(PRectangle rc, XYPOSITION /*cornerSize*/, FillS
 	for (int x = static_cast<int>(std::max(rc.left, clip.left)), y = static_cast<int>(rc.top - 1);
 		x < rc.right; x++) {
 		attr_t attrs = mvwinch(win, y, x) & A_ATTRIBUTES;
-		short pair = PAIR_NUMBER(attrs), fore = COLOR_WHITE, unused;
-		if (pair > 0) pair_content(pair, &fore, &unused);
-		mvwchgat(win, y, x, 1, attrs, PAIR_NUMBER(Colors::Pair(Colors::Find(fore), fill)), nullptr);
+		short pair = PAIR_NUMBER(attrs), fore = COLOR_WHITE, back = COLOR_BLACK;
+		if (pair > 0) pair_content(pair, &fore, &back);
+		if (fill.GetAlpha())
+			mvwchgat(win, y, x, 1, attrs, PAIR_NUMBER(Colors::Pair(Colors::Find(fore), fill)), nullptr);
+		else
+			mvwchgat(win, y, x, 1, attrs | A_UNDERLINE, PAIR_NUMBER(Colors::Pair(fill, Colors::Find(back))), nullptr);
 	}
 }
 
```

**Comment by rhaberkorn** [comment]
> Alternatively, an alpha value of 0 could cause underlines and coloring with INDIC_STRAIGHTBOX:

I am using this patch now in SciTECO. Would you like a PR?

**Comment by orbitalquark** [comment]
Thanks for the suggestions, but I am not in favor of using tricks like these to implement new behavior that is not self-documenting.

**Comment by rhaberkorn** [comment]
I do really need underlined indicators and would like to avoid maintaining patches on top of Scinterm.

So I've looked into INDIC_PLAIN and INDIC_STRIKE. They are calling SurfaceImpl::FillRectangle(). Unfortunately, I doubt these calls can be reliably kept apart and detected as underlines or strike-throughs. (We'd even have A_STRIKETHROUGH on PDCurses...) Perhaps it could be made to work if we'd touch Scintilla as well and change the FillRectangle() calls into LineDraw() or add an DrawUnderline() method or something.

Here however is an implementation of INDIC_SQUIGGLELOW to render it with A_UNDERLINE:

```cpp
// Used for wrap markers, margin markers and by INDIC_SQUIGGLE[LOW].
void SurfaceImpl::PolyLine(const Point *pts, size_t npts, Stroke stroke) {
	// Filter out everything except INDIC_SQUIGGLELOW
	for (int i = 1; i < npts; i++)
		if (pts[i].y != pts[0].y && pts[i].y != pts[0].y+1) return;

	int y = static_cast<int>(pts[0].y - 1);
	int x = static_cast<int>(pts[0].x);

	do {
		attr_t attrs = mvwinch(win, y, x) & A_ATTRIBUTES;
		short pair = PAIR_NUMBER(attrs), fore = COLOR_WHITE, back = COLOR_BLACK;
		if (pair > 0) pair_content(pair, &fore, &back);
		mvwchgat(win, y, x, 1, attrs | A_UNDERLINE, PAIR_NUMBER(Colors::Pair(stroke.colour, Colors::Find(back))), nullptr);
	} while (++x < static_cast<int>(pts[npts-1].x));
}
```

It will not currently work with INDIC_SQUIGGLE. On PDCurses there is also PDC_set_line_color(), which could perhaps be used to apply the stroke colour without changing the underlying text's color.
Considering that squiggles are sort of similar to underlines, perhaps you can accept this solution.

**Comment by rhaberkorn** [comment]
Here's the corresponding branch in my Scinterm fork: https://github.com/rhaberkorn/scinterm/tree/indic-squigglelow-underline

**Comment by orbitalquark** [comment]
I have implemented something similar in https://github.com/orbitalquark/scinterm/commit/8a0e888a87e760eec05224b93fdd8ea3c3be40f1. `PolyLine()` will only be called by squiggle indicators. All other calls are in Scintilla drawing routines Scinterm overrides. I also fixed drawing with `INDIC_SQUIGGLE`, which was not respecting the drawing clip.

**Comment by rhaberkorn** [comment]
> I have implemented something similar in [8a0e888](https://github.com/orbitalquark/scinterm/commit/8a0e888a87e760eec05224b93fdd8ea3c3be40f1).

Awesome. Thank you very much, now I can release SciTECO. Have a great new year!
